test/test.js

```diff
@@ -169,5 +169,42 @@ describe('Routes', () => {
       assert.strictEqual(res.status, 400)
       assert.strictEqual(await res.text(), 'string .walletAddress required')
     })
+    it('ignores duplicate submissions', async () => {
+      const createRequest = await fetch(
+        `${spark}/retrievals`,
+        { method: 'POST' }
+      )
+      const { id: retrievalId } = await createRequest.json()
+      {
+        const updateRequest = await fetch(
+          `${spark}/retrievals/${retrievalId}`,
+          {
+            method: 'PATCH',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify({ success: true, walletAddress })
+          }
+        )
+        assert.strictEqual(updateRequest.status, 200)
+      }
+      {
+        const updateRequest = await fetch(
+          `${spark}/retrievals/${retrievalId}`,
+          {
+            method: 'PATCH',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify({ success: false, walletAddress })
+          }
+        )
+        assert.strictEqual(updateRequest.status, 404)
```

Comment:
I think the status 404 may complicate troubleshooting errors, because we cannot tell whether there is no such retrieval job with `retrievalId` or whether the job exists but the result was already submitted.

I am proposing to return [400 Bad Request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) or [409 Conflict](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409) and include a descriptive error message in the response body.

It's fine to make this change in a follow-up PR.

Reply:
The update fails if we find no row with given id or without a success state. Currently we have no way to tell (without changing the query) whether it failed for the one reason or the other. Can you think of a simple way to change this?

I'm going to merge for now, let's continue discussing and I'll follow up with a PR for this

Reply:
> Currently we have no way to tell (without changing the query) whether it failed for the one reason or the other. Can you think of a simple way to change this?

I think we will need two queries, unless we can find some advanced SQL feature that will give us the behaviour we need.

Reply:
We can still make it atomic.
1. The first query finds the row by id. If the row does not exist, we return 404.
2. The second query updates the row if no result was recorded yet and returns a flag to indicate whether the query updated something or nothing.

There would be a race condition of somebody deleted the row between the first and the second query, in which case we would incorrectly say that the result was already reported instead of 404 not found. However, we are not deleting rows, so this situation is not going to happen.

Reply:
On the second thought, maybe we can do this the other way around and optimise for the happy path:

1. Run your current query. If it succeeds then all is good.
2. If we don't update anything, then run the second query to see if the retrieval id is invalid or we already have a result recorded.

WDYT?

Reply:
Ah nice I like that optimization for the happy path! Will follow up :)

Reply:
https://github.com/filecoin-station/spark-api/pull/17